### PR TITLE
Improve comprehension line break beheavior

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR improves the Black compatibility when it comes to breaking comprehensions. 

We want to avoid line breaks before the target and `in` whenever possible. Furthermore, `if X is not None` should be grouped together, similar to other binary like expressions

## Test Plan

`cargo test`

### CODE REVIEW

**Comment by MichaReiser** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5596** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5596" target="_blank">Graphite</a> 
    * **PR #5608** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5608" target="_blank">Graphite</a> 
      * **PR #5609** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5609" target="_blank">Graphite</a> 
        * **PR #5643** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5643" target="_blank">Graphite</a> 
          * **PR #5644** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5644" target="_blank">Graphite</a> 
            * **PR #5661** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5661" target="_blank">Graphite</a> 
              * **PR #5680** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5680" target="_blank">Graphite</a>  👈
        * **PR #5615** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5615" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5680?utm_source=stack-comment).

File: crates/ruff_python_formatter/src/other/comprehension.rs [hunk 1]
```diff
@@ -18,33 +30,40 @@ impl FormatNodeRule<Comprehension> for FormatComprehension {
...
             f,
             [
                 text("for"),
                 trailing_comments(before_target_comments),
                 group(&format_args!(
-                    soft_line_break_or_space(),
+                    Spacer(target),
                     target.format(),
-                    soft_line_break_or_space(),
+                    in_spacer,
```
Comment (konstin) [inline crates/ruff_python_formatter/src/other/comprehension.rs:62]: would the macros rules allow inlining the `in_space` as if expression?

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@py_37__python37.py.snap [hunk 2]
```diff
@@ -100,10 +90,9 @@ async def func():
     if test:
         out_batched = [
             i
-            async for
-            i
-            in
-            aitertools._async_map(self.async_inc, arange(8), batch_size=3)
+            async for i in aitertools._async_map(
```
Comment (konstin) [inline crates/ruff_python_formatter/tests/snapshots/black_compatibility@py_37__python37.py.snap:93]: that looks much nicer

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__fmtskip5.py.snap [hunk 3]
```diff
@@ -21,13 +21,12 @@ else:
...
 +++ Ruff
-@@ -1,7 +1,8 @@
+@@ -1,7 +1,7 @@
  a, b, c = 3, 4, 5
  if (
      a == 3
 -    and b    != 9  # fmt: skip
-+    and b
-+    != 9  # fmt: skip
++    and b != 9  # fmt: skip
```
Comment (konstin) [inline crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__fmtskip5.py.snap:29]: this, too

File: crates/ruff_python_formatter/tests/snapshots/format@statement__list_comp.py.snap [hunk 4]
```diff
@@ -80,6 +87,21 @@ input_file: crates/ruff_python_formatter/resources/test/fixtures/ruff/statement/
...
+
+[
+    aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+    + bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb
+    + [dddddddddddddddddd, eeeeeeeeeeeeeeeeeee]
+    for (
+        ccccccccccccccccccccccccccccccccccccccc,
+        ddddddddddddddddddd,
+        [eeeeeeeeeeeeeeeeeeeeee, fffffffffffffffffffffffff],
+    ) in eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeffffffffffffffffffffffffggggggggggggggggggggghhhhhhhhhhhhhhothermoreeand_even_moreddddddddddddddddddddd
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_formatter/tests/snapshots/format@statement__list_comp.py.snap:99]: yep i think that black remove tuple parentheses here same as `for` statements. i'd expect this is the same case for if-else expressions being analogous to `if` statements

**Review by konstin — APPROVED** [review]

**Comment by MichaReiser** [comment]
@MichaReiser merged this pull request with [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5680).

**Pull request merged**